dll/win32/shell32/CCopyToMenu.cpp

```diff
@@ -0,0 +1,258 @@
+#include "precomp.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+
+static HRESULT
+_GetCidlFromDataObject(IDataObject *pDataObject, CIDA** ppcida)
+{
+    static CLIPFORMAT s_cfHIDA = 0;
+    if (s_cfHIDA == 0)
+    {
+        s_cfHIDA = (CLIPFORMAT)RegisterClipboardFormatW(CFSTR_SHELLIDLIST);
+    }
+
+    FORMATETC fmt = { s_cfHIDA, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+    STGMEDIUM medium;
+
+    HRESULT hr = pDataObject->GetData(&fmt, &medium);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    LPVOID lpSrc = GlobalLock(medium.hGlobal);
+    SIZE_T cbSize = GlobalSize(medium.hGlobal);
+
+    *ppcida = (CIDA *)::CoTaskMemAlloc(cbSize);
+    if (*ppcida)
+    {
+        memcpy(*ppcida, lpSrc, cbSize);
+        hr = S_OK;
+    }
+    else
+    {
+        hr = E_FAIL;
+    }
+    ReleaseStgMedium(&medium);
+    return hr;
+}
+
+CCopyToMenu::CCopyToMenu() :
+    m_idCmdFirst(0),
+    m_idCmdLast(0),
+    m_idCmdCopyTo(-1)
+{
+}
+
+CCopyToMenu::~CCopyToMenu()
+{
+}
+
+static int CALLBACK
+BrowseCallbackProc(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
+{
+    CCopyToMenu *this_ =
+        reinterpret_cast<CCopyToMenu *>(GetWindowLongPtr(hwnd, GWLP_USERDATA));
+    LPCITEMIDLIST pidl;
+    WCHAR szPath[MAX_PATH];
+
+    switch (uMsg)
+    {
+        case BFFM_INITIALIZED:
+            SetWindowLongPtr(hwnd, GWLP_USERDATA, lpData);
+            this_ = reinterpret_cast<CCopyToMenu *>(lpData);
+
+            // Select initial directory
+            SendMessageW(hwnd, BFFM_SETSELECTION, FALSE,
+                reinterpret_cast<LPARAM>(static_cast<LPCITEMIDLIST>(this_->m_pidlFolder)));
+            break;
+
+        case BFFM_SELCHANGED:
+            pidl = reinterpret_cast<LPCITEMIDLIST>(lParam);
+            if (SHGetPathFromIDListW(pidl, szPath) && PathFileExistsW(szPath) &&
+                !ILIsEqual(pidl, this_->m_pidlFolder))
+            {
+                SendMessageW(hwnd, BFFM_ENABLEOK, 0, TRUE);
+            }
+            else
+            {
+                SendMessageW(hwnd, BFFM_ENABLEOK, 0, FALSE);
+            }
+            break;
+    }
+
+    return FALSE;
+}
+
+HRESULT CCopyToMenu::DoRealCopy(LPCMINVOKECOMMANDINFO lpici, LPCITEMIDLIST pidl)
+{
+    CComHeapPtr<CIDA> pCIDA;
+    HRESULT hr = _GetCidlFromDataObject(m_pDataObject, &pCIDA);
+    if (FAILED(hr))
+        return hr;
+
+    PCUIDLIST_ABSOLUTE pidlParent = HIDA_GetPIDLFolder(pCIDA);
+    if (!pidlParent)
+        return E_FAIL;
+
+    CStringW strFiles;
+    WCHAR szPath[MAX_PATH];
+    for (UINT n = 0; n < pCIDA->cidl; ++n)
+    {
+        PCUIDLIST_RELATIVE pidlRelative = HIDA_GetPIDLItem(pCIDA, n);
+        if (!pidlRelative)
+            continue;
+
+        PIDLIST_ABSOLUTE pidlCombine = ILCombine(pidlParent, pidlRelative);
+        if (!pidl)
+            return E_FAIL;
+
+        SHGetPathFromIDListW(pidlCombine, szPath);
+        ILFree(pidlCombine);
+
+        if (PathFileExistsW(szPath))
```

Comment:
Technically it does not have to be a real path, but we have this problem everywhere in our shell...

Reply:
OK.